src/metabase/csv.clj

```diff
@@ -0,0 +1,101 @@
+(ns metabase.csv
+  (:require
+   [clojure.data.csv :as csv]
+   [clojure.java.io :as io]
+   [clojure.set :as set]
+   [clojure.string :as str]
+   [medley.core :as m]
+   [metabase.search.util :as search-util]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [trs]]))
+
+;;           text
+;;            |
+;;            |
+;;       varchar_255
+;;            |
+;;            |
+;;          float
+;;            |
+;;            |
+;;           int
+;;            |
+;;            |
+;;         boolean
+(derive ::boolean     ::int)
+(derive ::int         ::float)
+(derive ::float       ::varchar_255)
+(derive ::varchar_255 ::text)
+
+(defn value->type
+  "The most-specific possible type for a given value. Possibilities are:
+    - ::boolean
+    - ::int
+    - ::float
+    - ::varchar_255
+    - ::text
+    - nil, in which case other functions are expected to replace it with ::text as the catch-all type
+
+  NB: There are currently the following gotchas:
+    1. ints/floats are assumed to have commas as separators and periods as decimal points
+    2. 0 and 1 are assumed to be booleans, not ints."
+  [value]
+  (cond
+    (str/blank? value)                                      nil
+    (re-matches #"(?i)true|t|yes|y|1|false|f|no|n|0" value) ::boolean
+    (re-matches #"-?[\d,]+"                          value) ::int
+    (re-matches #"-?[\d,]*\.\d+"                     value) ::float
+    (re-matches #".{1,255}"                          value) ::varchar_255
+    :else                                                   ::text))
+
+(defn- row->types
+  [column-names row]
+  (->> row
+       (map vector column-names)
+       (map (fn [[column-name value]] [column-name (and value (value->type (search-util/normalize value)))]))
+       (into {})))
+
+(defn coalesce
+  "Returns the 'parent' type (the most general)."
+  [type-a type-b]
+  (cond
+    (nil? type-a)        type-b
+    (nil? type-b)        type-a
+    (isa? type-a type-b) type-b
+    (isa? type-b type-a) type-a
+    :else
+    (let [common-ancestors (set/intersection (ancestors type-a) (ancestors type-b))]
```

Comment:
I don't think this is well-defined.

`(coalesce ::boolean ::boolean)` is obviously boolean. But the logic will be
Get the set of ancestors and intersect them, and then take the first element of the set. Whic is undefined. And since boolean is at the bottom, 

```clojure
csv=> (ancestors ::boolean)
#{:metabase.csv/text
  :metabase.csv/float
  :metabase.csv/int
  :metabase.csv/varchar_255}
csv=> (set/intersection (ancestors ::boolean) (ancestors ::boolean))
#{:metabase.csv/text
  :metabase.csv/float
  :metabase.csv/int
  :metabase.csv/varchar_255}
```
It is taking the first of the set of all types. Which is undefined. It will give an answer, but just any one will do.

Reply:
Boolean `isa?` boolean, so it'll hit an earlier branch and return boolean.

That said, the ancestors nonsense was from when I dug myself into hole with diamond inheritance that no longer exists, so I can just remove that logic entirely and `throw`.